### PR TITLE
Demote a warning to debug message, if C compiler is not there

### LINKED ISSUE #44177: Warning on libc not found for non-existing compiler
### Steps to reproduce

Using a `compilers.yaml` with compiler definitions for two different systems, with the non-host definitions in a path that does not exist:
```yaml
compilers:
- compiler:
    spec: gcc@=9.4.0
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    flags: {}
    operating_system: ubuntu20.04
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
- compiler:
    spec: gcc@=9.3.0
    paths:
      cc: /some/path/gcc
      cxx: /some/path/g++
      f77: /some/path/gfortran
      fc: /some/path/gfortran
    flags: {}
    operating_system: ubuntu22.04
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
```
Concretitization always warns about not finding `libc` for the non-existing compiler:
```console
$ spack spec zlib-ng
```

### Error message

```console
==> Warning: cannot detect libc from gcc@=9.3.0. The compiler will not be used during concretization.
Input spec
--------------------------------
 -   zlib-ng

Concretized
--------------------------------
 -   zlib-ng@2.1.6%gcc@9.4.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=linux-ubuntu20.04-icelake
 -       ^gcc-runtime@9.4.0%gcc@9.4.0 build_system=generic arch=linux-ubuntu20.04-icelake
[e]      ^glibc@2.31%gcc@9.4.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -       ^gmake@4.4.1%gcc@9.4.0~guile build_system=generic arch=linux-ubuntu20.04-icelake

```

### Information on your system

* **Spack:** 0.23.0.dev0 (7f3dd38ccc106142846f29f76b928fa58e0e3164)
* **Python:** 3.11.5
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

@becker33

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by alalazo**
fixes #44177

### CODE REVIEW

**Comment by alalazo** [comment]
This will leave a warning when a C compiler is in the expected path, but a libc cannot be detected. We could filter further using the OS, if we pre-compute which OS is compatible with one of the OSes supported by the platform. Not sure it's worth implementing that...

**Comment by haampie** [comment]
Alternatively check `c.compiler_verbose_output`, cause existence of a path does not imply the compiler runs.

File: lib/spack/spack/solver/asp.py [hunk 1]
```diff
@@ -2975,6 +2979,13 @@ class CompilerParser:
     def __init__(self, configuration) -> None:
         self.compilers: Set[KnownCompiler] = set()
         for c in all_compilers_in_config(configuration):
+            if using_libc_compatibility() and not c_compiler_runs(c):
```
Comment (haampie) [inline lib/spack/spack/solver/asp.py:2982]: can you merge the two branches? next one is pretty much identical. I think it's more helpful to tell users that there is a reason to have a "working" compiler, namely to detect libc. If there's no `compiler_verbose_output` the message can be extended with "couldn't run the compiler" or "couldn't get the link line from the compiler" or something.

Comment (haampie) [reply lib/spack/spack/solver/asp.py:2982]: oh, right, i forgot that this was about debug message vs warnings... nevermind then.

**Review by haampie — APPROVED** [review]

**Pull request merged**